[go1.17] Update to go1.17

justaugustus opened the pull request:
#### What type of PR is this?

/kind feature
/area dependency

#### What this PR does / why we need it:

[go1.17] Update to go1.17
Tracking issue: https://github.com/kubernetes/release/issues/2169

Signed-off-by: Stephen Augustus <foo@auggie.dev>

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Kubernetes is now built using go1.17
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

justaugustus committed:
[go1.17] Update to go1.17

Signed-off-by: Stephen Augustus <foo@auggie.dev>

dims commented:
is it that time already... wow! rc2!

justaugustus commented:
> is it that time already... wow! rc2!

Hehe, it really does fly by, doesn't it?

dims commented:
hmmm go.mod/sum files in vendor/ are wiped?

justaugustus commented:
For my own paranoia:
/test pull-kubernetes-dependencies

liggitt commented on build/dependencies.yaml:
```diff
@@ -101,30 +101,36 @@ dependencies:
+    version: 1.16.0
+    refPaths:
+    - path: hack/lib/golang.sh
+      match: minimum_go_version=go([0-9]+\.[0-9]+)
```
hmm... this is going to cause issues with people running gofmt and update-vendor and getting wildly different results locally with go 1.16.x

liggitt commented on go.mod:
```diff
@@ -132,10 +249,20 @@ require (
+	sigs.k8s.io/kustomize/kyaml v0.11.0 // indirect
+)
+
+require (
```
the three require blocks is interesting... I think it's related to https://tip.golang.org/doc/go1.17#go-command:

> Because the number of explicit requirements may be substantially larger in an expanded Go 1.17 go.mod file, the newly-added requirements on indirect dependencies in a go 1.17 module are maintained in a separate require block from the block containing direct dependencies.

but the end result here is... weird... there aren't cleanly separated "direct requires" and "indirect requires" blocks... I'm pulling this branch down and trying to see what's going on

liggitt commented:
oof... the go.mod file output by go1.17 is really confusing our depstats:

```
-	"directDependencies": 161,
+	"directDependencies": 229,
```

liggitt commented:
while we deal with the fallout of the `go mod tidy` changes in go1.17, let's leave the go.mod files at `go 1.16` to unblock the rest of the PR

liggitt commented on go.mod:
```diff
@@ -6,7 +6,7 @@
 module k8s.io/kubernetes
 
-go 1.16
+go 1.17
```
keep this at `go 1.16` for now, we can spawn a follow up issue to decide how to deal with the `tidy` changes in 1.17

liggitt commented on go.mod:
```diff
@@ -443,20 +570,20 @@ replace (
 	golang.org/x/image => golang.org/x/image v0.0.0-20190802002840-cff245a6509b
 	golang.org/x/lint => golang.org/x/lint v0.0.0-20210508222113-6edffad5e616
 	golang.org/x/mobile => golang.org/x/mobile v0.0.0-20201217150744-e6ae53a27f4f
 	golang.org/x/mod => golang.org/x/mod v0.4.2
-	golang.org/x/net => golang.org/x/net v0.0.0-20210525063256-abc453219eb5
-	golang.org/x/oauth2 => golang.org/x/oauth2 v0.0.0-20210514164344-f6687ab2804c
+	golang.org/x/net => golang.org/x/net v0.0.0-20210805182204-aaa1db679c0d
+	golang.org/x/oauth2 => golang.org/x/oauth2 v0.0.0-20210810183815-faf39c7919d5
 	golang.org/x/sync => golang.org/x/sync v0.0.0-20210220032951-036812b2e83c
-	golang.org/x/sys => golang.org/x/sys v0.0.0-20210616094352-59db8d763f22
-	golang.org/x/term => golang.org/x/term v0.0.0-20210220032956-6a3ed077a48d
+	golang.org/x/sys => golang.org/x/sys v0.0.0-20210809222454-d867a43fc93e
+	golang.org/x/term => golang.org/x/term v0.0.0-20210615171337-6886f2dfbf5b
 	golang.org/x/text => golang.org/x/text v0.3.6
 	golang.org/x/time => golang.org/x/time v0.0.0-20210723032227-1f47c861a9ac
-	golang.org/x/tools => golang.org/x/tools v0.1.2
+	golang.org/x/tools => golang.org/x/tools v0.1.6-0.20210811045132-92b2fbe72692
```
revert this file to match `master` except for these intentional bumps, then run `update-vendor.sh` again... the diff should be *much* smaller

liggitt commented on hack/tools/go.mod:
```diff
@@ -1,6 +1,6 @@
 module k8s.io/kubernetes/hack/tools
 
-go 1.16
+go 1.17
```
revert to match `master` and keep at go 1.16 for now

justaugustus committed:
[go1.17] Bump golang.org/x/... dependencies

hack/pin-dependency.sh golang.org/x/crypto master
hack/pin-dependency.sh golang.org/x/net master
hack/pin-dependency.sh golang.org/x/oauth2 master
hack/pin-dependency.sh golang.org/x/sync master
hack/pin-dependency.sh golang.org/x/sys master
hack/pin-dependency.sh golang.org/x/term master
hack/pin-dependency.sh golang.org/x/time master
hack/pin-dependency.sh golang.org/x/tools master

Signed-off-by: Stephen Augustus <foo@auggie.dev>

justaugustus committed:
generated: Run hack/lint-dependencies.sh and hack/update-vendor.sh

Signed-off-by: Stephen Augustus <foo@auggie.dev>

caesarxuchao commented:
/remove-sig api-machinery

justaugustus committed:
generated: Run hack/update-gofmt.sh

Signed-off-by: Stephen Augustus <foo@auggie.dev>

justaugustus commented on build/tools.go:
```diff
@@ -1,3 +1,4 @@
+//go:build tools
 // +build tools
```
Looks like the build tag style is slightly different.
Will probably need to fix these up individually...

justaugustus replied:
Possibly, but not in our current invocation of it: https://github.com/kubernetes/kubernetes/blob/2584a7c0bb27a5029c3ac24b496a20da58edca35/hack/update-gofmt.sh#L46

`//go:build tools` seems like it should be equivalent to `// +build tools`, but the new tag prepended instead of replaced.

liggitt replied:
> //go:build tools seems like it should be equivalent to // +build tools

I'm confused… according to https://tip.golang.org/doc/go1.17#gofmt and https://go.googlesource.com/proposal/+/master/design/draft-gobuild.md they are equivalent, right?

liggitt replied:
prepending is the expected behavior… the old style is maintained for compatibility with older go versions

justaugustus replied:
Okay, perfect (re: prepend vs replace).
(I'm AFK, so didn't have the relnotes in front of me.)

liggitt commented:
verify boilerplate needs update to tolerate the new build tag, something like:

```diff
diff --git a/hack/boilerplate/boilerplate.py b/hack/boilerplate/boilerplate.py
index 5a6d3177eb1..4d105c054eb 100755
--- a/hack/boilerplate/boilerplate.py
+++ b/hack/boilerplate/boilerplate.py
@@ -218,7 +218,7 @@ def get_regexs():
     regexs["date"] = re.compile(get_dates())
     # strip // +build \n\n build constraints
     regexs["go_build_constraints"] = re.compile(
-        r"^(// \+build.*\n)+\n", re.MULTILINE)
+        r"^(//(go:build| \+build).*\n)+\n", re.MULTILINE)
     # strip #!.* from scripts
     regexs["shebang"] = re.compile(r"^(#!.*\n)\n*", re.MULTILINE)
     # Search for generated files
```

and it looks like one dependency.yaml failure:
```
ERRO /home/prow/go/src/k8s.io/kubernetes/build/dependencies.yaml indicates that golang: upstream version should be at version 1.17rc2, but the following files didn't match: cluster/images/etcd/Makefile 
```

other than that, I *think* the rest of the CI failures are due to CI running with go1.16 trying to verify files generated/formatted with go1.17rc2

do we have canary versions of pull-kubernetes-dependencies / pull-kubernetes-verify / pull-kubernetes-integration / pull-kubernetes-unit running with go1.17rc2 so we can have confidence all CI is actually green on go1.17 prior to merge?

justaugustus committed:
dependencies.yaml: Alpha-sort reference paths

Signed-off-by: Stephen Augustus <foo@auggie.dev>

justaugustus commented:
> do we have canary versions of pull-kubernetes-dependencies / pull-kubernetes-verify / pull-kubernetes-integration / pull-kubernetes-unit running with go1.17rc2 so we can have confidence all CI is actually green on go1.17 prior to merge?

Let's find out:
/test ls

If not, we'll get them created!

liggitt commented:
/test pull-kubernetes-dependencies-go-canary
/test pull-kubernetes-integration-go-canary
/test pull-kubernetes-verify-go-canary

That looks super promising (if the go-canary jobs are using go1.17rc2). unit tests are the only other job that seems likely to use CI-specific go

justaugustus committed:
dependencies.yaml: Track Golang version for upstream etcd releases

Signed-off-by: Stephen Augustus <foo@auggie.dev>

justaugustus committed:
hack/boilerplate: Tolerate new build tag format (`//go:build`)

Signed-off-by: Stephen Augustus <foo@auggie.dev>

justaugustus commented:
/test pull-kubernetes-dependencies-go-canary
/test pull-kubernetes-integration-go-canary
/test pull-kubernetes-verify-go-canary

liggitt commented:
huh... go-canary verify job hit this:

```
verify: govet
# k8s.io/kubernetes/test/images/agnhost/pause
test/images/agnhost/pause/pause.go:41:2: misuse of unbuffered os.Signal channel as argument to signal.Notify 
test/images/agnhost/pause/pause.go:42:2: misuse of unbuffered os.Signal channel as argument to signal.Notify

# k8s.io/kubernetes/test/e2e_node/runner/remote 
test/e2e_node/runner/remote/run_remote.go:209:2: misuse of unbuffered os.Signal channel as argument to signal.Notify 

# k8s.io/kubernetes/test/images/agnhost/nettest 
test/images/agnhost/nettest/nettest.go:220:3: misuse of unbuffered os.Signal channel as argument to signal.Notify
```

I wonder if those are real problems... that's fun

justaugustus commented:
/test pull-kubernetes-verify
/test pull-kubernetes-e2e-gci-gce-ipvs
/test pull-kubernetes-node-kubelet-serial

One more run of the canaries:
/test pull-kubernetes-dependencies-go-canary
/test pull-kubernetes-integration-go-canary
/test pull-kubernetes-verify-go-canary

justaugustus commented:
> > do we have canary versions of pull-kubernetes-dependencies / pull-kubernetes-verify / pull-kubernetes-integration / pull-kubernetes-unit running with go1.17rc2 so we can have confidence all CI is actually green on go1.17 prior to merge?
> 
> Let's find out:
> /test ls
> 
> If not, we'll get them created!

Sent some updates to Golang canary presubmits here: https://github.com/kubernetes/test-infra/pull/23247

justaugustus commented:
/test pull-kubernetes-unit-go-canary

justaugustus commented:
A few "fun" ones from [`pull-kubernetes-unit-go-canary`](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/103692/pull-kubernetes-unit-go-canary/1426265427960401920):

```console
 === FAIL: pkg/apis/certificates/validation Test_validateCertificateSigningRequestOptions/status.certificate,_non-ASN1_data (0.00s)
    validation_test.go:1070: missing expected strict error: status.certificate: Invalid value: "<certificate data>": asn1: structure error: sequence tag mismatch
    validation_test.go:1073: unexpected strict error: status.certificate: Invalid value: "<certificate data>": x509: invalid RDNSequence
    --- FAIL: Test_validateCertificateSigningRequestOptions/status.certificate,_non-ASN1_data (0.00s)
=== FAIL: pkg/apis/certificates/validation Test_validateCertificateSigningRequestOptions (0.12s)
=== FAIL: plugin/pkg/admission/serviceaccount TestAddImagePullSecrets (0.00s)
W0813 19:51:24.879892   71471 mutation_detector.go:53] Mutation detector is enabled, this will result in memory leakage.
    admission_test.go:760: 
        	Error Trace:	admission_test.go:760
        	Error:      	Not equal: 
        	            	expected: []v1.LocalObjectReference([]v1.LocalObjectReference{v1.LocalObjectReference{Name:"foo"}, v1.LocalObjectReference{Name:"bar"}})
        	            	actual  : []core.LocalObjectReference([]core.LocalObjectReference{core.LocalObjectReference{Name:"foo"}, core.LocalObjectReference{Name:"bar"}})
        	Test:       	TestAddImagePullSecrets
        	Messages:   	expected [{foo} {bar}], got [{foo} {bar}]
=== FAIL: vendor/k8s.io/apimachinery/pkg/util/validation TestIsValidIP (0.00s)
    validation_test.go:340: expected true for "1.1.1.01": [must be a valid IP address, (e.g. 10.9.8.7 or 2001:db8::ffff)]
=== FAIL: vendor/k8s.io/apimachinery/pkg/util/validation TestIsValidIPv4Address (0.00s)
    validation_test.go:368: expected "1.1.1.01" to be valid IPv4 address: [[]: Invalid value: "1.1.1.01": must be a valid IPv4 address]
=== FAIL: vendor/k8s.io/apiserver/pkg/util/webhook TestKubeConfigFile/cluster_with_invalid_CA_certificate_-_parse_error (0.01s)
    webhook_test.go:309: cluster with invalid CA certificate - parse error: unexpected error message to match:
          Expected: unable to load root certificates: failed to parse certificate: asn1: syntax error: data truncated
          Actual:   unable to load root certificates: failed to parse certificate: x509: malformed certificate
    --- FAIL: TestKubeConfigFile/cluster_with_invalid_CA_certificate_-_parse_error (0.01s)
=== FAIL: vendor/k8s.io/apiserver/pkg/util/webhook TestKubeConfigFile (0.13s) 
```

liggitt commented:
the certificate ones look like a hard-coded error message changed, that'll be an easy test fix

the valid IP ones are catching go breaking backcompat since we haven't switched to our pinned ip parsing yet (https://github.com/kubernetes/kubernetes/issues/100895)

this is the only one that looks suspicious:
```
=== FAIL: plugin/pkg/admission/serviceaccount TestAddImagePullSecrets (0.00s)
W0813 19:51:24.879892   71471 mutation_detector.go:53] Mutation detector is enabled, this will result in memory leakage.
    admission_test.go:760: 
        	Error Trace:	admission_test.go:760
        	Error:      	Not equal: 
        	            	expected: []v1.LocalObjectReference([]v1.LocalObjectReference{v1.LocalObjectReference{Name:"foo"}, v1.LocalObjectReference{Name:"bar"}})
        	            	actual  : []core.LocalObjectReference([]core.LocalObjectReference{core.LocalObjectReference{Name:"foo"}, core.LocalObjectReference{Name:"bar"}})
        	Test:       	TestAddImagePullSecrets
        	Messages:   	expected [{foo} {bar}], got [{foo} {bar}]
```

liggitt commented:
certificate message unit tests updated in https://github.com/kubernetes/kubernetes/pull/104416

fedebongio commented:
/remove-sig api-machinery

liggitt commented:
> this is the only one that looks suspicious:
>
> === FAIL: plugin/pkg/admission/serviceaccount TestAddImagePullSecrets (0.00s)

verified this is a behavior change in go reflect IsConvertible behavior... opening an upstream issue

liggitt commented:
@justaugustus the PRs fixing the unit test failures have merged... on a rebase this *should* be good to go now

liggitt commented:
oops, missed the signal channel thing - https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/103692/pull-kubernetes-verify-go-canary/1426243405683363840

justaugustus commented:
> @justaugustus the PRs fixing the unit test failures have merged... on a rebase this _should_ be good to go now
> oops, missed the signal channel thing - https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/103692/pull-kubernetes-verify-go-canary/1426243405683363840
> 
> fixed in #104489

@liggitt -- Thanks for the help, as always! I've rebased, but leaving the `WIP` on until I fix up the official images in https://github.com/kubernetes/release/pull/2211.

justaugustus commented:
/test pull-kubernetes-dependencies-go-canary
/test pull-kubernetes-integration-go-canary
/test pull-kubernetes-unit-go-canary
/test pull-kubernetes-verify-go-canary

justaugustus commented:
`pull-kubernetes-verify` failure is expected.
/test pull-kubernetes-e2e-gce-storage-slow
/test pull-kubernetes-e2e-gce-ubuntu-containerd

justaugustus commented:
/test pull-kubernetes-e2e-gce-storage-slow

justaugustus commented:
/test pull-kubernetes-e2e-gce-storage-slow
/test pull-kubernetes-e2e-gce-ubuntu-containerd

justaugustus commented:
/test pull-kubernetes-node-kubelet-serial

aojea commented:
```
go/format
The Source and Node functions now synchronize //go:build lines with // +build lines. If a file only has // +build lines, they will be moved to the appropriate location in the file, and matching //go:build lines will be added. Otherwise, // +build lines will be overwritten based on any existing //go:build lines. For more information, see https://golang.org/design/draft-gobuild.
```
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/103692/pull-kubernetes-verify/1428844005319249920

liggitt commented:
Yeah, that version is still running go1.16. https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/103692/pull-kubernetes-verify-go-canary/1428850292065046528 is running with go1.17 and passed

justaugustus commented:
/test pull-kubernetes-dependencies-go-canary
/test pull-kubernetes-integration-go-canary
/test pull-kubernetes-unit-go-canary
/test pull-kubernetes-verify-go-canary

justaugustus commented:
Ready for review!
cc: @kubernetes/release-engineering @liggitt @dims

liggitt commented on build/dependencies.yaml:
```diff
@@ -85,32 +85,43 @@ dependencies:
+      match: GOLANG_VERSION=\d+.\d+(alpha|beta|rc)?\.?\d+
 
-  # Golang pre-releases are denoted as `1.y<pre-release stage>.z`
+  # TODO(releng): Bump after:
+  #               1. go1.17 is released
+  #               2. initial go1.17 PR has merged
+  - name: "golang: minimum supported version"
+    version: 1.16.0
```
go ahead and drop this stanza... we want CI to require go1.17 before this PR merges

liggitt commented on build/dependencies.yaml:
```diff
@@ -85,32 +85,43 @@ dependencies:
+    version: 1.17
     refPaths:
     - path: build/build-image/cross/VERSION
-    - path: hack/lib/golang.sh
-      match: minimum_go_version=go([0-9]+\.[0-9]+)
```
restore this stanza, it will pass the canary CI jobs and fail the blocking ones. once we have green canaries, we'll switch the blocking ones to go1.17 at which point this will become the only PR that can merge

liggitt replied:
and I guess update golang.sh to require go1.17 now

liggitt commented:
changes lgtm, I think the order of operations are:
1. update dependencies.yaml and golang.sh to require go1.17
2. get green CI runs on canary jobs
3. switch CI blocking jobs to go1.17
4. retest/merge this PR

liggitt commented:
/test pull-kubernetes-dependencies-go-canary
/test pull-kubernetes-integration-go-canary
/test pull-kubernetes-unit-go-canary
/test pull-kubernetes-verify-go-canary

justaugustus commented:
Status:
1. DONE - update dependencies.yaml and golang.sh to require go1.17
2. [WIP](https://github.com/kubernetes/kubernetes/pull/103692#issuecomment-904932119) - get green CI runs on canary jobs
3. [WIP](https://github.com/kubernetes/test-infra/pull/23354) - switch CI blocking jobs to go1.17
4. retest/merge this PR

justaugustus commented:
Email sent to k-dev: https://groups.google.com/g/kubernetes-dev/c/5kiB72M8518
ref: https://github.com/kubernetes/test-infra/pull/23358

liggitt commented:
/lgtm
/approve
/retest

liggitt commented:
/skip

liggitt commented:
labeling to bump to the top of the merge queue since CI requires 1.17 now

dims commented:
/approve

(in principle)

liggitt commented:
Looks like I beat the CI 1.17 bump

/retest

justaugustus commented:
Here we gooooooo!
/pony party
/honk

justaugustus commented:
The kind jobs were still on the older image version:
/retest

justaugustus commented:
/test pull-kubernetes-e2e-kind-ipv6

justaugustus commented:
/skip

justaugustus commented:
/test pull-kubernetes-unit

```console
=== RUN   TestMultithreadedThrottling
    throttle_test.go:71: Expected it to take at least 1 second to get 100 tokens, took 985.174006ms
--- FAIL: TestMultithreadedThrottling (0.99s)
```

sanposhiho commented:
After this PR, it seems that all CI `pull-kubernetes-conformance-kind-ipv6-parallel` fails with the below error.

```
Detected go version: go version go1.16.7 linux/amd64.
Kubernetes requires go1.17.0 or greater.
Please install go1.17.0 or later.
```

https://prow.k8s.io/job-history/gs/kubernetes-jenkins/pr-logs/directory/pull-kubernetes-conformance-kind-ipv6-parallel

Do we have to do something?

justaugustus commented:
> After this PR, it seems that all CI `pull-kubernetes-conformance-kind-ipv6-parallel` fails with the below error.
> 
> ```
> Detected go version: go version go1.16.7 linux/amd64.
> Kubernetes requires go1.17.0 or greater.
> Please install go1.17.0 or later.
> ```
> 
> https://prow.k8s.io/job-history/gs/kubernetes-jenkins/pr-logs/directory/pull-kubernetes-conformance-kind-ipv6-parallel
> 
> Do we have to do something?

@sanposhiho -- The kind presubmit versions were bumped in https://github.com/kubernetes/test-infra/pull/23358, but the PRs you're watching may be reporting the results of jobs using the old image. Issuing a `/retest` command (on a line by itself) so trigger a new test run with the correct image.

(We had to do the same for this PR a few times.)

sanposhiho commented:
Thanks, I'll give it a try 🏃

sanposhiho commented:
@justaugustus 
I'm not familiar with test-infra repo, but isn't this CI task still goes on the old image(gcr.io/k8s-testimages/krte:v20210808-1eaeec7-master)? 

https://github.com/kubernetes/test-infra/blob/master/config/jobs/kubernetes/sig-testing/conformance-e2e.yaml#L64

kerthcet commented:
> @justaugustus
> I'm not familiar with test-infra repo, but isn't this CI task still configured to go on the old image(gcr.io/k8s-testimages/krte:v20210808-1eaeec7-master)?
> 
> https://github.com/kubernetes/test-infra/blob/master/config/jobs/kubernetes/sig-testing/conformance-e2e.yaml#L64
> 
> It seems some people also run test many times but no test passes.
> https://prow.k8s.io/job-history/gs/kubernetes-jenkins/pr-logs/directory/pull-kubernetes-conformance-kind-ipv6-parallel

```
+ make all WHAT=cmd/kubectl test/e2e/e2e.test vendor/github.com/onsi/ginkgo/ginkgo

Detected go version: go version go1.16.7 linux/amd64.
Kubernetes requires go1.17.0 or greater.
Please install go1.17.0 or later.

!!! [0825 05:32:34] Call tree:
!!! [0825 05:32:34]  1: hack/run-in-gopath.sh:31 kube::golang::setup_env(...)

Detected go version: go version go1.16.7 linux/amd64.
Kubernetes requires go1.17.0 or greater.
Please install go1.17.0 or later.

!!! [0825 05:32:34] Call tree:
!!! [0825 05:32:34]  1: /home/prow/go/src/k8s.io/kubernetes/hack/lib/golang.sh:793 kube::golang::setup_env(...)
!!! [0825 05:32:34]  2: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
!!! [0825 05:32:34] Call tree:
!!! [0825 05:32:34]  1: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
make[1]: *** [Makefile.generated_files:208: _output/bin/prerelease-lifecycle-gen] Error 1
make: *** [Makefile:552: generated_files] Error 2
wrapper.sh] [TEST] Test Command exit code: 2
```

kerthcet commented:
> > After this PR, it seems that all CI `pull-kubernetes-conformance-kind-ipv6-parallel` fails with the below error.
> > ```
> > Detected go version: go version go1.16.7 linux/amd64.
> > Kubernetes requires go1.17.0 or greater.
> > Please install go1.17.0 or later.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > https://prow.k8s.io/job-history/gs/kubernetes-jenkins/pr-logs/directory/pull-kubernetes-conformance-kind-ipv6-parallel
> > Do we have to do something?
> 
> @sanposhiho -- The kind presubmit versions were bumped in [kubernetes/test-infra#23358](https://github.com/kubernetes/test-infra/pull/23358), but the PRs you're watching may be reporting the results of jobs using the old image. Issuing a `/retest` command (on a line by itself) should trigger a new test run with the correct image.
> 
> (We had to do the same for this PR a few times.)

`/retest` not work I think. I checked the latest branch, it seems like all the go version is upgraded to `1.17`. keep tracing.

kerthcet commented:
related:  https://github.com/kubernetes/test-infra/pull/23364

aojea commented:
it is failing because the images of the CI weren't updated, the bot has to update the new images

https://github.com/kubernetes/test-infra/pull/23143

aojea commented:
The change is here in the queue https://github.com/kubernetes/test-infra/pull/23349

medyagh commented:
This PR breaks kubernetes/minikube's automation that bumps the golang version to match kubernetes/kubernetes 

is there any reason that 1.23.0 Prefix was added to this image?

liggitt commented:
> is there any reason that 1.23.0 Prefix was added to this image?

because there are more inputs to the image than just the golang version (like the protobuf generators), and we don't always want to bump those on release branches

justaugustus commented:
> because there are more inputs to the image than just the golang version (like the protobuf generators), and we don't always want to bump those on release branches

+1. Explained in detail in https://github.com/kubernetes/release/pull/2211:

> * images: Enforce Debian codenames for Golang-based images
>   When there are multiple image builds in flight both upstream and downstream,
>   we can run into situations where a new Debian version becomes the default for
>   image builds, which can have unintended side-effects across release branches.
>   Here we use explicit pairings of Golang/Debian versions to retrieve images
>   Example: `FROM golang:1.16.7-buster`
> 
> * kube-cross: Use OS codenames to construct clearer versions for images
>   Uses the following nomenclature:
>   `v<kubernetes-major>-go<go-major>-<os-codename>.<revision>`
>   Example: `v1.23.0-go1.17-buster.0`

medyagh commented:
thanks for the explain @liggitt  @justaugustus  now it makes sense !

justaugustus commented:
> thanks for the explain @liggitt @justaugustus now it makes sense !

Happy to help, @medyagh!
